### DISCUSSION TITLE
Hotswap plate offset and tilted switches

### DISCUSSION THREAD

**Discussion by ivanmgribeiro** [joshreve/dactyl-keyboard#91] [General]
Hi,
I printed a keyboard using this repo with the HS_NUB setting and noticed that the switches do not sit level when put in. I've also noticed that this has been mentioned in other discussions (https://github.com/joshreve/dactyl-keyboard/discussions/15 and https://github.com/joshreve/dactyl-keyboard/discussions/66).
I had a look and I think the hotswap plate needs to be offset in the y direction by 0.5mm. At the moment, the hotswap plate is too far forward in the switch hole, so the cylinder that sticks out at the bottom of the switches is pushed forward which tilts the switch, and also means that the pins are not quite aligned to the hotswap socket.
I've made the following change: (https://github.com/ivanmgribeiro/dactyl-keyboard/commit/07bbf4306ead5cded0d997188a1207eb7d90dd2b), which fixes the issue. I haven't confirmed that the issue exists with the HS_UNDERCUT, but it exists in both the HS_NUB and HS_NOTCH versions. I've only tested the fix with HS_NOTCH but it should also fix the issue in the HS_NUB version.

Before:
![Before](https://user-images.githubusercontent.com/38882872/170756653-16cf0a84-25c5-44c0-8a6c-6ec1962f7bf0.jpg)

After:
![IMG_2022-05-26-14-48-20-121](https://user-images.githubusercontent.com/38882872/170757119-ee341569-3d04-4dd0-b5e1-775609f87a65.jpg)

Thanks for your work on this repo, and I hope this is useful. I thought it would be better to create a new discussion rather than use the old ones to make it easier for people to find this later on.

**Comment by bosd** [comment]
Did you also have the issue that the plate needs to be rotated?

Can you make a picture of the bottom, how the switch fits the nub like in:
https://github.com/joshreve/dactyl-keyboard/discussions/66#discussioncomment-2811080

I'm dependent on a external print service, just wanna make sure me next print/design is correct. :)

**Reply by ivanmgribeiro** [reply]
I didn't have any problems with needing to rotate switches/baseplates.
I'm not sure what issue you're referring to, I couldn't tell from the pictures you posted before what the problem was.
Also I'd highly recommend using HS_NOTCH rather than HS_NUB.

Left is original print (no offset, labelled "Before" above), right is new print (with offset, labelled "After" above). Let me know if you need pictures of a different angle
![switches-under](https://user-images.githubusercontent.com/38882872/171088427-b6d6c0c3-3f08-4654-916c-22a35e42bc57.jpg)

**Comment by bosd** [comment]
Thanks for the detailed pictures!
We are talking about related issues :smile: 
I'm even not at the point I could use the bottom plate. As the problem in my print lies to the fact that the switches don't fit inside the wholes.

On a different note, It is odd that we both have issues with the fitting of these switches.
In my case I downloaded the stl's from another repo. The build from that person was perfect.

**Trying to solve the Kaihl switches fitting mystery..**
Now I look closer at your latest pictures. I've noticed that the bottom of your switches look different.
The box of the switch seems to have a different design.
Mine looks like https://user-images.githubusercontent.com/16755871/143526000-4dcd2b77-4ffd-4638-a9f8-65b9c673f74b.png
Could it be that the dimensions of the box switches itself are slightlty different from series?
(Sadly I don't have other switches lying around)
I'm using the Kahil Silent Box Browns and Silent Box Pinks..

**Reply by ivanmgribeiro** [reply]
The switches that I use are Cherry MX Silent Red.
I think all switches have a standard design, so I don't think that's the problem.

When you say "the switches don't fit inside the holes", do you mean that the switch sits like in this image: (https://user-images.githubusercontent.com/11499387/170005276-bbaf19ad-2bb8-4aa0-b9c3-6709504f6cd3.jpg) or that it does not fit at all?

Can you send some pictures of a switch inside the hole that is broken in one of your prints (https://user-images.githubusercontent.com/11499387/170015825-b900b090-929d-41c1-8e76-c54384fd7d0c.png)? Ideally pictures with it rotated and not rotated.

BTW if you could reply to this comment (not at the bottom of the page) that will make the discussion easier to follow.

**Reply by ivanmgribeiro** [reply]
I also have switches that look a little more similar to yours:
![bobau4](https://user-images.githubusercontent.com/38882872/171114441-2c382908-5423-45b3-ba3c-a2e15720feb0.jpg)
These are Gazzew Boba U4 switches. They fit identically to the Cherry switches.

**Reply by ivanmgribeiro** [reply]
Also: the stl files in the other repo were generated using this repo - the author of the other repo mentions this in the readme.

**Reply by bosd** [reply]
> When you say "the switches don't fit inside the holes", do you mean that the switch sits like in this image: (https://user-images.githubusercontent.com/11499387/170005276-bbaf19ad-2bb8-4aa0-b9c3-6709504f6cd3.jpg) or that it does not fit at all?

I mean indeed that they sit on top as in the picture your are referring to.

**Reply by bosd** [reply]
> Can you send some pictures of a switch inside the hole that is broken in one of your prints (https://user-images.githubusercontent.com/11499387/170015825-b900b090-929d-41c1-8e76-c54384fd7d0c.png)? Ideally pictures with it rotated and not rotated.

![IMG_20220531_103642](https://user-images.githubusercontent.com/11499387/171131601-c335403d-90d5-492f-82fa-41a48853b00d.jpg)

Here is a picture of the switch, inserted in the socket without baseplate. It is inserted in the original direction.
There is no way I can get it to fit in the hole. The Nub design is making the hole to narrow.

**Reply by bosd** [reply]
> Can you send some pictures of a switch inside the hole that is broken in one of your prints (https://user-images.githubusercontent.com/11499387/170015825-b900b090-929d-41c1-8e76-c54384fd7d0c.png)? Ideally pictures with it rotated and not rotated.

Here is a picture with the switch rotated. (If there was a baseplate the solder pins would hit the baseplate.)
The switch fits. Kinda hard to describe, but it is not really locked/clicked into place. As I would expect to be the purpose of the nub design.
You still need a bit of force to pull the switch out. But was Expecting a snappier fit so one has to use a bit more force to pull the switch out.
 
![IMG_20220531_104354_1](https://user-images.githubusercontent.com/11499387/171133663-cb4da3af-e461-448b-82e7-41510048801d.jpg)

**Reply by ivanmgribeiro** [reply]
OK just to be clear: When you put the switch in the hole without a baseplate, does it fit in properly in the hole (like in the "After" picture), or does it sit at an angle (like in https://user-images.githubusercontent.com/11499387/170005276-bbaf19ad-2bb8-4aa0-b9c3-6709504f6cd3.jpg)? Also, is the "Nub" the bit in blue that I have drawn on this image:
![switch-in](https://user-images.githubusercontent.com/38882872/171133718-dfa6d054-907a-4d88-ac97-ab993db9af48.jpg)

If it is then I think the switches you have might actually be different. The ones that I've seen all have an indentation (like the one in https://user-images.githubusercontent.com/16755871/143526000-4dcd2b77-4ffd-4638-a9f8-65b9c673f74b.png below the red circle).

Sorry these questions might seem basic, but I don't have the same issues that you're having so I'm just trying to understand the issue better.

**Reply by bosd** [reply]
> OK just to be clear: When you put the switch in the hole without a baseplate, does it fit in properly in the hole

When using the hole without plate, inserting the switch at a 90 degr rotation. It fits..
See picture for clarification
![IMG_20220531_110612](https://user-images.githubusercontent.com/11499387/171137375-ca6dbbb2-5084-4d9c-800c-72b238babcf5.jpg)

Glad that you ask all these questions so we can get to the bottom of this :smile:

**Reply by ivanmgribeiro** [reply]
That does indeed look like the correct fit. I guess if it is not rotated then it does not fit as well as that?

**Reply by bosd** [reply]
indeed, inserting like the solder pins match the bottom plate does not fit at al and gives this result:
https://user-images.githubusercontent.com/11499387/170005276-bbaf19ad-2bb8-4aa0-b9c3-6709504f6cd3.jpg

**Comment by bosd** [comment]
I've found this https://github.com/tshort/dactyl-keyboard/pull/36
That seems a perfect fit for the hot swaps, and propably introduces the snappy fixing I was expecting.

**Reply by ivanmgribeiro** [reply]
I think this is what the HS_NOTCH option in this repo does. The test prints I did with this option are snappy.

**Reply by ivanmgribeiro** [reply]
I think for your print what you want is to implement the offset that I mention at the top of this discussion, and also use HS_NOTCH instead of HS_NUB. That should give you a good design.
I plan on printing a 4x6 dactyl manuform with these settings this weekend. I don't have any kailh-style switches so I won't be able to test how those fit but I can let you know if I have any problems.

**Reply by bosd** [reply]
> I think for your print what you want is to implement the offset that I mention at the top of this discussion, and also use HS_NOTCH instead of HS_NUB. That should give you a good design.

I think you're right. Had a quick peek in freecad with the generated files. Looks ok.
FYI, While generating these files I still had some odd behaviour of the docker. Cannot pin point it yet. but sometimes I get the error that the `hot_swap_plate.step` could not be found.
One time, I was able to overcome this by rebuilding the docker containers.
Another go, I could not overcome this problem. Checked permissions but the issue seemed path related. 
So I've modded line 59 in dactyl_manuform.py to:
https://github.com/tshort/dactyl-keyboard/pull/36

I'm very interested to hear feedback from your build.

FYI, I've got some more challenges.
Maybe it's the printjob... Maybe the design.
The oled Clip just won't snap into place.
![IMG_20220531_112911(1)](https://user-images.githubusercontent.com/11499387/171210068-a17f8d2d-36d4-4e49-a580-5bf27cc6fd38.jpg)


![IMG_20220531_113004(1)](https://user-images.githubusercontent.com/11499387/171210104-51695f96-d279-4671-abd6-c57cc92ef4ac.jpg)
I've tried to place it upside down in the slot.. It has too much stress on the holder. Imo the holder should fit thight but not so that it bends the clip.

**Reply by ivanmgribeiro** [reply]
I think you've not linked the change to `dactyl_manuform.py` correctly - I don't see anything in that pull request that looks very relevant.
Regarding the OLED clip, I can't help with that because that's not something I've looked at and I don't plan on using it.

**Reply by ivanmgribeiro** [reply]
I got the prints finished over the weekend. The switch holes are great, all the switches click in properly. In case you want to have a look, the changes I made to the repo for my print are the last three commits here: (https://github.com/ivanmgribeiro/dactyl-keyboard/tree/a472dd06c793aa39a79aca41ca980697ea50b8c2).

I wouldn't recommend printing the exact settings that I have above, I removed the bottom row which had only 2 keys in it and that ended up curving the inside of the keyboard quite a bit more than I'd wanted to. I also had to change the attachment between the thumb cluster and the rest of the keyboard because the hotswap plate of the top-left key in the left thumb cluster was sticking out of the model.

In terms of the individual switch holes themselves they're perfect, all of them are like the "After" picture in the original post.

**Reply by bosd** [reply]
Thanks for your feedback!! Hope you enjoy your project!
